core/domain/stats_jobs_one_off.py

```diff
@@ -16,16 +16,212 @@
+                yield (item.exploration_id,
+                       {
+                           'type': 'state_hit',
+                           'version': item.exploration_version,
+                           'state': item.state_name,
+                           'session_id': item.session_id
+                       })
+            elif isinstance(item, stats_models.SolutionHitEventLogEntryModel):
+                yield (item.exp_id,
+                       {
+                           'type': 'solution_hit',
+                           'version': item.exp_version,
+                           'state': item.state_name,
+                           'session_id': item.session_id
+                       })
+            elif isinstance(
+                    item,
+                    stats_models.ExplorationActualStartEventLogEntryModel):
+                yield (item.exp_id,
+                       {
+                           'type': 'exploration_actual_start',
+                           'version': item.exp_version,
+                           'state': item.state_name
+                       })
+            elif isinstance(
+                    item, stats_models.CompleteExplorationEventLogEntryModel):
+                yield (item.exploration_id,
+                       {
+                           'type': 'complete_exploration',
+                           'version': item.exploration_version,
+                           'state': item.state_name
+                       })
+
+
+    @staticmethod
+    def reduce(key, values):
+        values = map(ast.literal_eval, values)
+        values = sorted(values, key=lambda x: x['version'])
+
+        for version, events in itertools.groupby(values,
+                                                 key=lambda x: x['version']):
+            if version == 1:
+                # Get a copy of the uncorrupted *_v1 statistics
+                old_stats = stats_services.get_exploration_stats_by_id(
+                    key, version)
+                for state_stats in old_stats.state_stats_mapping.values():
+                    state_stats.total_answers_count_v2 = 0
+                    state_stats.useful_feedback_count_v2 = 0
+                    state_stats.total_hit_count_v2 = 0
+                    state_stats.first_hit_count_v2 = 0
+                    state_stats.num_times_solution_viewed_v2 = 0
+                    state_stats.num_completions_v2 = 0
+                RecomputeStatistics._recompute_statistics(
+                    key, version, events, old_stats)
+            else:
+                change_list = exp_models.ExplorationCommitLogEntryModel.get(
+                    'exploration-%s-%s' % (key, version)).commit_cmds
+
+                old_exp_version = version - 1
+                exploration_stats = stats_services.get_exploration_stats_by_id(
+                    key, old_exp_version)
+                if exploration_stats is None:
```

Comment:
This ideally has no way of happening right? Not saying its wrong to guard against it anyway.

Reply:
Yes, the way I have it now with the single GET op removes the need for this I think.